### PR TITLE
Verify the bootstrap client cert before using it

### PR DESCRIPTION
**Pull request by smarterclayton**
Before the bootstrap client is used, check a number of conditions that
ensure it can be safely loaded by the server. If any of those conditions
are invalid, re-bootstrap the node. This is primarily to force
bootstrapping without human intervention when a certificate is expired,
but also handles partial file corruption.

### CODE REVIEW

**Comment by smarterclayton** [comment]
Second commit detects expired CSR status.certificate.  I'm not sure whether we want to give permission for CSRs to delete their own CSRs, but the current logic for server certificate rotation will wedge if the CSR is expired and the node is rebooted.

We should probably either iterate a sequence (add `-%d` to the end and increment) or try to delete.  We probably need the signing controller to purge certificates that have expired for short rotation to be practical.

Was testing with 10 minute expirations.

**Comment by liggitt** [comment]
> current logic for server certificate rotation will wedge if the CSR is expired and the node is rebooted

Really? New private key and name should be used for each new CSR… node should be able to just abandon an expired one

**Comment by smarterclayton** [comment]
Nothing is regenerating the private key for client or server certs that I see.

**Comment by smarterclayton** [comment]
Ok, I saw we changed the server cert to use Start() always.  The code here is still correct - an expired CSR is not compatible.

**Comment by smarterclayton** [comment]
I found the issue.  Before we start the cert manager we have to perform the initial bootstrap (using the bootstrap kubeconfig).  If we find a valid key on disk, we never regenerate it.  Because the key never changes, after the CSR expires the kubelet will be unable to start (the old key will be used with the same client info, which means we get the same CSR name).

`bootstrap.LoadClientCert` should delete the private key if it fails to get a new CSR for any reason, probably.

**Comment by liggitt** [comment]
> bootstrap.LoadClientCert should delete the private key if it fails to get a new CSR for any reason, probably.

@mikedanese wasn't this to be re-entrant if the CSR wasn't satisfied because the manual approval went long or the controller was backed up, and the kubelet timed out and restarted?

**Comment by smarterclayton** [comment]
In this case the CSR is expired, so we are out of luck without deleting the
CSR or picking a new name.

**Comment by smarterclayton** [comment]
Added a third commit - it may be too aggressive for cleanup (always), so I can move it later if necessary (if request CSR exits out)

**Comment by smarterclayton** [comment]
Found a new problem - if a client cert is used for creating CSRs and expires, or if a token is used for creating CSRs and it expires, or if the auth mechanism is revoked, the kubelet will just spin forever trying to bootstrap and anything using a client will effectively be dead.  This is exacerbated when the admin sets an aggressive CSR window - the odds of a kubelet pausing / missing a loop / having connection problems can then put the kubelet into a dead state where it never proceeds.

I *think* in general there needs to be some time based guarantee that ensures that the kubelet starts to crash loop when successive rotations fail.  As long as a crash looping kubelet gets at least one chance to request a new cert using whatever credentials are on disk that is sufficient.  It should not be possible for a kubelet to go zombie.

Scenarios for getting a "cert retrieval client" when rotation is on:

1. kubelet has a bootstrap kube config
2. kubelet has no bootstrap kube config, but flags that tell it to use a kubeconfig on disk
3. kubelet has a previous client cert on disk that the cert store tries to use

Failure modes for the "cert retrieval client":

1. bootstrap kube config is expired or has permission revoked
2. flags for kube config point to expired or revoked tokens
3. admin revokes permission on bootstrap kubeconfig
4. the automated mechanism that puts the config on the node dies

Time when the "cert retrieval client" expires:

1. Before the cert manager starts
2. After the cert manager loads it but before the first request
3. After the first request

Types of errors:

1. 403 - cert expired
2. 403 - master rotated its certs
3. 403 - token expired
4. 403 - RBAC removed permission
5. 403 - code upgrade
6. 404 - cert rotation endpoint is off
7. 500 - master is crashing
8. timeout - master is unreachable

Suggestions:

1. provide an upper bound in time for a kubelet cert manager to fail continuously before it exits
2. check the "cert retrieval client" (whether from bootstrap, store, or kube config) and if it is expired, exit quickly

We assume that in a heavy rotation environment clock skew is unacceptable (skew must be << rotation interval).

**Comment by liggitt** [comment]
Cases 1,2,3 would be 401, not 403

**Comment by smarterclayton** [comment]
The last commit (which exits after it can no longer fetch a valid cert) may be too aggressive.  It bounds how long a master outage can go on before the cluster goes from coasting (no API updates, but continue to maintain existing pods) to decaying (because the kubelet has no state).  Eventually checkpointing may reduce that impact, but the maximum downtime of an api server bounds how long the cluster goes before it degrades.

**Comment by smarterclayton** [comment]
Proposal - if the server is still responsive, restart.  If the server isn't returning a 401, stay up because availability may be compromised.  Basically, stay a zombie as long as it's better than the alternative.

File: pkg/kubelet/certificate/bootstrap/bootstrap.go [hunk 1]
```diff
@@ -82,20 +94,20 @@ func LoadClientCert(kubeconfigPath string, bootstrapPath string, certDir string,
 	if err != nil {
 		return fmt.Errorf("unable to build bootstrap client cert path: %v", err)
 	}
+	defer func() {
```
Comment (mikedanese) [inline pkg/kubelet/certificate/bootstrap/bootstrap.go]: why is this moved above RequestNodeCertificate as well? Does that write anything?

**Review by mikedanese — COMMENTED** [review]
This all looks reasonable to me.

File: pkg/kubelet/certificate/bootstrap/bootstrap.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/kubelet/certificate/bootstrap/bootstrap.go]: To avoid leaving a cert on disk if we got a key but failed during req node certificate.  The cert is useless without the key (which we delete above).  This makes the flow a lot more predictable - if we fail to attain a cert, delete key and cert and try again.

**Comment by smarterclayton** [comment]
I'm going to add one more commit that makes the node go into zombie mode if the server is unreachable when we try to request a cert, but die quickly if the cert is up (to give us another chance to bootstrap).  That's assuming the bootstrap lifetime is longer than the cert lifetime, but if the cert lifetime is much greater than bootstrap lifetime this is less of an issue.

**Comment by smarterclayton** [comment]
I stuck one commit on at the end - instead of doing request client cert and waiting for an hour, the cert manager now does shorter waits and tracks whether the server is responding at all.  If we get a definitive response from the server, then the cert manager should not exit when the client cert expires and instead keep on in zombie mode.

1. kubelet starts, gets bootstrap client cert
2. gives bootstrap client cert to cert manager, which requests a new node cert
3. cert expires because the server didn't sign in time
4. cert manager knows that it can't sign a new cert, and it sees 401 or valid errors from the server, so it assumes that it can restart and either bootstrap, or signal a crash loop

The alternate for 4 is that the server isn't able to even contact the server because it gets connection refused, timeout, or a generic proxy error, and so the transport code allows the kubelet to keep limping along after its connection has expired in order to not leave its pods completely unmanaged while the masters are down.

**Comment by mikedanese** [comment]
I haven't reviewed the last commit yet. It's on my todo list.

**Comment by smarterclayton** [comment]
Updated to react to the change to move certificate manager into client-go - kubelet/util/csr is generic and moved with it.  The node part is effectively generic and can be an example to users.
